sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -664,9 +715,18 @@ public WriteFn(Write<T> spec) {
 
       @Setup
       public void setup() throws Exception {
-        dataSource = spec.getDataSourceConfiguration().buildDatasource();
+        if (spec.getDataSourceFactory() != null) {
+          dataSource = spec.getDataSourceFactory().create();
+        } else {
+          dataSource = spec.getDataSourceConfiguration().buildDatasource();
+        }
+      }
+
+      @StartBundle
+      public void startBundle() throws Exception {
```

Comment:
What does a user need to do to ensure that `getConnection()` in the line below uses a connection pool, rather than establishes a new connection every time? I agree that the current PR is the right thing to do in general, but I'm worried about people getting a big performance hit if they use their existing code unchanged.

Reply:
It's up to the user to provide a `DataSource` supporting pooling. You are right: if he uses a "non-pooled" `DataSource`, then a new connection will be created for each bundle.
Should I add a note in the javadoc about that ? Or I can add an option to let the user decides when he wants to create the connection: at `Setup` or at `Bundle`. WDYT ?

Reply:
Is it possible to provide a pooled DataSource by default?

Reply:
Right now, the user provides his own datasource (it could be a generic with DBCP, or a Oracle pooled datasource, or a MySQL pooled datasource). So, using `DataSource`, it could be pooled or not. We can "force" the usage of DBCP (wrapping the provided `DataSource` with a DBCP pool). Thoughts ?

Reply:
Maybe, I can "request" a `ConnectionPoolDataSource` (https://docs.oracle.com/javase/7/docs/api/javax/sql/ConnectionPoolDataSource.html) to the user. WDYT ?

Reply:
Maybe I can check if the user provides a `ConnectionPoolDataSource`, then, I set the connection per bundle, else, I set the connection at `Setup` (as it is today). Thoughts ?

Reply:
I think wrapping with a pool (of size 1, because it's only used by a single thread) is a reasonable option.

However, note that if the user didn't provide a DataSource or a DataSourceFactory, then we are creating a BasicDataSource which is already pooling - in that case we don't need to wrap.

Please note in both cases that by default the connection pool will behave exactly as bad as the current implementation, keeping the idle connections open forever. To make it actually close idle connections, see https://stackoverflow.com/questions/38685092/dbcp2-when-are-idle-connections-removed-from-the-pool